### PR TITLE
Refactor scatter/gather ops to use the new cost based threadpool abstractions.

### PR DESCRIPTION
**Pull request by pranavsharma**
**Description**: Refactor scatter/gather ops to use the new cost based threadpool abstractions. Also removed unnecessary usages of openmp. Tested on all models in the perf dashboard. For non-openmp builds it seemed to perform better than the master for most models. The openmp builds should've no impact.

**Motivation and Context**
Consistent usage of threapool APIs across all ops that account for openmp and Eigen in one place.

### CODE REVIEW

**Comment by pranavsharma** [comment]
I've one more change coming up in mlas/lib/threading.cpp file. Prefer to separate this from mlas change for easy debugging in case there are any perf regressions.

File: onnxruntime/core/providers/cpu/tensor/gather.cc [hunk 1]
```diff
@@ -93,7 +90,13 @@ Status GatherCopyData(const Tensor* indices_tensor, const uint8_t* src_base, uin
     } else {
       memcpy(dst_base + dst_offset, src_base + src_offset, block_size);
     }
-  }
+  };
+  concurrency::ThreadPool::TryParallelFor(tp, M * N, static_cast<double>(block_size),
+                                          [lambda](ptrdiff_t first, ptrdiff_t last) {
+                                            for (int index = static_cast<int>(first), end = static_cast<int>(last); index < end; ++index) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/tensor/gather.cc:96]: &lambda?

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc [hunk 2]
```diff
@@ -81,77 +81,83 @@ Status ScatterNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) co
...
-#endif
-  for (int64_t i = 0; i < offset_count; ++i) {
+  auto lambda = [&](int64_t i) {
     for (int64_t j = 0; j < last_indice_dimension; ++j) {
       auto indice = *(indice_offset + i * last_indice_dimension + j);
       if (indice < 0 || indice >= input_shape[j]) {
         err_indice = indice;
       }
       p.element_offsets[i] += indice * element_counts[j];
     }
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/tensor/scatter_nd.cc:113]: Does this need to be parallelized? I suspect there's not enough work and doing so will be slower.

File: onnxruntime/core/providers/cpu/tensor/gather.cc (hunk 1 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/tensor/gather.cc:96]: Yeah, probably not. I'll change it. Got carried over from the lstm code.

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc (hunk 2 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc:113]: Didn't seem to make much of a difference. Removed it.

**Pull request merged**